### ISSUE TITLE
IDLC dds_topic_descriptor_t intializer causes struct ending up .data section

### ISSUE THREAD

**Issue by PetervdPerk-NXP** [eclipse-cyclonedds/cyclonedds#1624]
I'm using IDLC on a embedded platform and noticed that my declerations of `dds_topic_descriptor_t` ends up in the .data sections. It seems this is due that the IDLC intialize code doesn't intitialize the `restrict_data_representation` field causing the compiler see the struct as potential mutable and therefore puts it in the .data section.

Since I'm working on a small embedded platform I want to make sure that if something is constant it ends up in the .rodata section.
Is there a reason that `restrict_data_representation` doesn't get initialized?

**Comment by dpotman** [comment]
Yes, that will work with [these](https://github.com/eclipse-cyclonedds/cyclonedds/compare/master...dpotman:topic-desc-ser-static?expand=1) changes, but I'm not really happy with the case where the type info and type mapping are dynamically generated in `ddsi_typebuilder.c`, that needs a cast to a non-const pointer for freeing the descriptor. But I think we cannot avoid that. @eboasson any suggestions?

**Comment by eboasson** [comment]
I would say that the intent has always been that these things can be stored in `.rodata` and that that's why the original bunch of fields use `const`.

Because of that choice, we already have the problem of having to cast away the `const` in some cases (e.g., with the type name), so we're not really introducing a new problem if we do this, we're merely making it consistent. That consistency and the fact that it once again can be put in `.rodata` are both improvements in my view.

In short, while I'm not *happy* to have those casts I don't see any good alternatives right now and I think that the benefits outweigh the cost.

**Comment by dpotman** [comment]
Thanks @eboasson, I've created PR https://github.com/eclipse-cyclonedds/cyclonedds/pull/1626 for these changes.
